shared/api/v3/course.ts

```diff
@@ -0,0 +1,185 @@
+}
+export const EnrollmentStatusEnum = z.nativeEnum(EnrollmentStatus);
+export type EnrollmentStatusEnum = z.infer<typeof EnrollmentStatusEnum>;
+
+export const Course = z.object({
+    courseCode: z.string(),
+    courseName: z.string(),
+    courseSortKey: z.string().array(),
+    courseMutualExclusionKey: z.string().array(),
+    courseDescription: z.string(),
+    courseInstructors: z.string().array(),
+    courseTerm: z.string(),
+    courseSchedule: Schedule.array(),
+    courseCredits: z.number().nonnegative(),
+    courseSeatsTotal: z.number().int().nonnegative(),
+    courseSeatsFilled: z.number().int().nonnegative(),
+    courseWaitlistLength: z.null(),
+    courseEnrollmentStatus: EnrollmentStatusEnum,
+    permCount: z.number().int().nonnegative(),
+});
+export type Course = z.infer<typeof Course>;
+
+export function courseFromV4Section(s: APIv4.Section): Course {
+    const sectionCodeString = APIv4.stringifySectionCode(s.identifier);
+    const courseCodeString = APIv4.stringifyCourseCode(s.course.code);
+
+    const startDate = stringifyCourseDate(s.startDate);
+    const endDate = stringifyCourseDate(s.endDate);
+
+    const { termCount, terms } = getSectionTermFractions(s.identifier);
+
+    return {
+        courseCode: sectionCodeString,
+        courseName: s.course.title,
+        courseSortKey: [sectionCodeString],
+        courseMutualExclusionKey: [courseCodeString],
+        courseDescription: s.course.description,
+        courseInstructors: s.instructors.map((inst) => inst.name),
+        courseTerm: stringifySectionTermCode(s.identifier),
+        courseSchedule: s.schedules.flatMap((schedule) => {
+            const days = stringifyWeekdayEnumArray(schedule.days);
+            const startTime = stringifyTime(schedule.startTime);
+            const endTime = stringifyTime(schedule.endTime);
+            return schedule.locations.map((location) => ({
+                scheduleDays: days,
+                scheduleStartTime: startTime,
+                scheduleEndTime: endTime,
+                scheduleStartDate: startDate,
+                scheduleEndDate: endDate,
+                scheduleTermCount: termCount,
+                scheduleTerms: terms,
+                scheduleLocation: location,
+            }));
+        }),
+        courseCredits: s.credits,
+        courseSeatsTotal: s.seatsTotal,
+        courseSeatsFilled: s.seatsFilled,
+        courseWaitlistLength: null,
+        courseEnrollmentStatus: convertStatusEnum(s.status),
+        permCount: s.permCount,
+    };
+}
+
+export function courseListFromV4SectionList(sections: APIv4.Section[]): {
+    terms: [];
+    courses: Course[];
+} {
+    return {
+        terms: [],
+        courses: sections.flatMap((section) => {
+            try {
+                const course = courseFromV4Section(section);
+                return [course];
+            } catch (e) {
+                console.error(e);
+                return [];
+            }
+        }),
```

Comment:
If you change this chunk to a `for-of` loop you don't need to return empty arrays, but simply pushing data to a pre-allocated array. This avoids multiple allocations (and given the size of our data, about 2.5k) of arrays that are later merged back into the resulting array each time this function is called. You can just do `continue` inside the loop to skip pushing broken data.